File: utils/XeThru_utils/xeX4Thru_software/ModuleConnector/ModuleConnector-unix-1/python35-x86_64-linux-gnu/pymoduleconnector/examples/X4M200_sleep_record.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" \example X4M200_sleep_record.py

#Target module: X4M200

#Introduction: This is an example of how to record respiration and movement data for sleep monitoring from X4M200.

#Command to run: "python X4M200_sleep_record.py -d com8" or "python3 X4M200_sleep_record.py -d com8"
                 change "com8" with your device name, using "--help" to see other options.
                 Using TCP server address as device name is also supported, e.g. 
                 "python X4M200_sleep_record.py -d tcp://192.168.1.169:3000".
"""

from __future__ import print_function
import sys
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from optparse import OptionParser
import numpy as np
from time import sleep
from pymoduleconnector import ModuleConnector
from pymoduleconnector import DataType
from pymoduleconnector import DataRecorder
from pymoduleconnector import RecordingOptions
import pymoduleconnector


def on_file_available(data_type, filename):
    print("new file available for data type: {}".format(data_type))
    print("  |- file: {}".format(filename))
    if data_type == DataType.BasebandApDataType:
        print("processing baseband ap data from file")
    elif data_type == DataType.SleepDataType:
        print("processing sleep data from file")


def on_meta_file_available(session_id, meta_filename):
    print("new meta file available for recording with id: {}".format(session_id))
    print("  |- file: {}".format(meta_filename))


def record_sleep (device_name):
    directory = "."

    # Reset module
    mc = ModuleConnector(device_name)
    x4m200 = mc.get_x4m200()
    x4m200.module_reset()
    mc.close()
    sleep(3)

    # Assume an X4M300/X4M200 module and try to enter XEP mode
    mc = ModuleConnector(device_name)
    x4m200 = mc.get_x4m200()
    recorder = mc.get_data_recorder()

    recorder = mc.get_data_recorder()
    recorder.subscribe_to_file_available(pymoduleconnector.AllDataTypes, on_file_available)
    recorder.subscribe_to_meta_file_available(on_meta_file_available)
    recorder.start_recording(DataType.BasebandApDataType | DataType.SleepDataType, directory)

    # Stop running application and set module in manual mode.
    try:
        x4m200.set_sensor_mode(0x01, 0) # Make sure no profile is running
    except RuntimeError:
        # Profile not running, OK
        pass

    # Load x4m200 respiration detection profile
    x4m200.load_profile(0x47fabeba)
    try:
        x4m200.set_sensor_mode(0x01,0)#RUN mode
    except RuntimeError:
        # Sensor already stopped, OK
        pass
    x4m200.set_output_control(0x610a3b00,1)
    while True:
        rdata = x4m200.read_message_respiration_sleep()
        print ("Frame: {} RPM: {} Distance: {} Movement Slow: {} Movement Fast: {}" .format(rdata.frame_counter, rdata.respiration_rate, rdata.distance, rdata.movement_slow, rdata.movement_fast))
        sleep(0.2)


def main():
    parser = OptionParser()
    parser.add_option(
        "-d",
        "--device",
        dest = "device_name",
        help = "device file to use",
        metavar = "FILE")

    (options, args) = parser.parse_args()

    if not options.device_name:
        print ("you have to specify device, e.g.: python record.py -d /dev/ttyACM0")
        sys.exit(1)

    record_sleep(options.device_name)


if __name__ == "__main__":
    main()

File: utils/XeThru_utils/xeX4Thru_software/ModuleConnector/ModuleConnector-rpi-1/examples/X4M200_sleep_record.cpp
/*
X4M200-recording example code
*/
#include "ModuleConnector.hpp"
#include "DataRecorder.hpp"
#include "functional"
#include "X4M200.hpp"
#include "iostream"
#include "unistd.h"
#include "stdio.h"
#include "signal.h"
#include "xtid.h"

using namespace XeThru;

volatile sig_atomic_t stop_recording;
void handle_sigint(int num)
{
    stop_recording = 1;
}

static void on_file_available(XeThru::DataType type, const std::string &filename)
{
    std::cout << "recorded file available for data type: "
              << DataRecorder::data_type_to_string(type) << std::endl;
    std::cout << "file: " << filename << std::endl;
}

static void on_meta_file_available(const std::string &session_id, const std::string &meta_filename)
{
    std::cout << "meta file available for recording with id: " << session_id << std::endl;
    std::cout << "file: " << meta_filename << std::endl;
}

int record(const std::string &device_name)
{
    using namespace XeThru;
    //Acquiring DataRecorder and ModuleConnector
    ModuleConnector mc(device_name, 0);
    DataRecorder &recorder = mc.get_data_recorder();

    const DataTypes data_types = BasebandApDataType | SleepDataType; //specifying data_types
    const std::string output_directory = "."; //setting directory to current

    //start recorder
    if (recorder.start_recording(data_types, output_directory) != 0) {
        //Start recording failed
        std::cout << "Failed to start recording" << std::endl;
        return 1;
    }

    // Get interface for physical device and enable data output

    {
        DataRecorder::FileAvailableCallback callback = std::bind(&on_file_available,
                                                                 std::placeholders::_1,
                                                                 std::placeholders::_2);
        recorder.subscribe_to_file_available(BasebandApDataType | SleepDataType, callback);
    }

    {
    DataRecorder::MetaFileAvailableCallback callback = std::bind(&on_meta_file_available,
                                                                 std::placeholders::_1,
                                                                 std::placeholders::_2);
    recorder.subscribe_to_meta_file_available(callback);
    }

    //configure and run X4M200
    X4M200 &x4m200 = mc.get_x4m200();
    x4m200.load_profile(XTS_ID_APP_RESPIRATION_2);
    x4m200.set_output_control(XTS_ID_RESPIRATION_DETECTIONLIST, XTID_OUTPUT_CONTROL_ENABLE);
    x4m200.set_sensor_mode(XTID_SM_RUN, 0);
    //Sleepdata variable
    XeThru::SleepData data;

    while (!stop_recording) {
        if (x4m200.read_message_respiration_sleep(&data)) {
            //X4M200 read failed
            std::cout << "Read message failed" << '\n';
            return 1;
        }
        //print output
        std::cout << "Counter " << data.frame_counter<< "  ";
        std::cout << "RPM " << data.respiration_rate << "  ";
        std::cout << "Movement Slow " << data.movement_slow << "  ";
        std::cout << "Movement Fast " << data.movement_fast << '\n';

        usleep(100);
    }
    return 0;
}

int main(int argc, char ** argv)
{
    if (argc < 2) {
        std::cout << "record <com port or device file>" << std::endl;
        return 1;
    }
    stop_recording = 0;
    signal(SIGINT, handle_sigint);
    return record(argv[1]);
}
